Allow for commit prefix

amitayk opened the pull request:
A new configuration option called "prefix" has been added. This option allows you to set a default prefix for each commit. It also supports regular expressions (regex) to match the prefix against the branch name.

Here is a simple example of using the prefix option:

oc config set prefix amyu98-commit:

This will result in:

amyu98-commit: Some lovely AI commit content

Below is an example of using a basic regex with the prefix option:

oc config set prefix /^([a-zA-Z]+-\d+)/

For the local branch "DEV-12345-commit-prefix", this will result in:

DEV-12345: Some lovely AI commit content

amitayk committed:
feat(api.ts): add support for prefix configuration to be added to the generated message
feat(config.ts): add prefix configuration option to be used in the generated message. Add validation to ensure prefix is not empty.

amitayk committed:
refactor(api.ts): improve readability by adding a space between prefix and message content
fix(config.ts): fix config validation for prefix to allow non-empty values

amitayk committed:
fix(api.ts): fix generatePrefix function to return undefined when config.prefix is undefined instead of falsy

amitayk committed:
feat(api.ts): add support for generating commit message prefix from regex pattern in git branch name
feat(api.ts): add support for generating commit message prefix from git branch name if regex pattern is not provided
feat(api.ts): add getCurrentGitBranch function to get current git branch name
feat(api.ts): add generatePrefixFromRegex function to generate prefix from regex pattern in git branch name

amyu98 commented:
issue: https://github.com/di-sukharev/opencommit/issues/152

takuya-o commented on src/api.ts:
```diff
@@ -94,4 +99,57 @@ export const getOpenCommitLatestVersion = async (): Promise<
+    return undefined;
+  }
+
+  return match[0];
```
The example regex had () such as /^([a-zA-Z]+-\d+)/. Is it expecting the whole match if the parentheses are missing, or the first parentheses if the parentheses are present?
```suggestion
  r return match.length > 1 ? match[1] : match[0];
```

takuya-o commented on package.json:
```diff
@@ -64,6 +64,7 @@
     "@dqbd/tiktoken": "^1.0.2",
     "axios": "^1.3.4",
     "chalk": "^5.2.0",
+    "child_process": "^1.0.2",
```
Other git command calls used execa.

amitayk committed:
Update src/api.ts

Improve generatePrefixFromRegex function to handle capturing groups

Co-authored-by: Takuya Ono <takuya-o@users.osdn.me>

amitayk committed:
refactor(api.ts): replace child_process.execSync with execaCommandSync to improve security and reliability
refactor(api.ts): simplify prefix generation logic by checking if prefix is defined and not empty before using it
fix(api.ts): fix conditional statement in generatePrefixFromRegex to return undefined when no match is found
fix(api.ts): catch error in getCurrentGitBranch and log it to console instead of throwing it

amyu98 commented:
@takuya-o Thanks, added suggestion and changed to execa (+ minor change in handling undefined prefix)

amitayk committed:
chore(package.json): remove unused "child_process" dependency to declutter dependencies list

snirbenyosef commented:
@amyu98 Amazing, waiting for it to be released.

takuya-o reviewed (commented):
Confirmed that it has changed to execa

czlowiek488 commented:
Any updates on this feature?
I really want to use this package but without this feature it is not usable for me 😞

AmitayKaufmamn commented:
@czlowiek488 Not quite sure who is supposed to push this towards merge. Do I need to do anything?

di-sukharev commented on src/commands/config.ts:
```diff
@@ -109,7 +110,17 @@ export const configValidators = {
+  [CONFIG_KEYS.prefix](value: any) {
+    validateConfig(
+      CONFIG_KEYS.prefix,
+      true,
```
true is not validating anything here :) we should validate the config input with a function or expression which return Boolean

amyu98 replied:
@di-sukharev Do you have an idea what can we validate?

di-sukharev replied:
you somehow need to confirm that the value is the exact format you expect, if your format is `^(*)` and not `/^(*)/` — make sure you don't have `/`s in the value, you may also do multiple validations like `!a && b && !c`

di-sukharev reviewed (commented):
Hi, sorry foe being late, have you tested the code?

To test you run: `npm run build && npm run start` inside the repo, so it runs actual build files and not `oc` which is installed globally

di-sukharev reviewed (commented):
Hi, sorry for being late :) Great work!

have you tested the code?

To test you run: `npm run build && npm run start` inside the repo, so it runs actual build files and not `oc` which is installed globally

and there are some conflicts to be fixed

amitayk committed:
Merge with master

tomtastico commented:
Any news on getting this merged? Would be awesome 🚀

di-sukharev commented:
@tomtastico i hope soon, we need to make `validateConfig()` actually validate something

amitayk committed:
Allow For Prefix: fix(config.ts): fix typo in OCO_PREFIX environment variable name
feat(config.ts): add validation for OCO_PREFIX environment variable to ensure it is a valid string or regex pattern

amitayk committed:
Allow For Prefix: refactor(api.ts): improve prefix validation by using regex instead of startsWith and endsWith methods

amitayk committed:
build

amitayk committed:
Allow For Prefix: refactor(api.ts): extract message prefixing logic to a separate function to improve readability and maintainability
feat(api.ts): add prefix to the message returned by getOpenCommitLatestVersion function to improve semantics and consistency with other functions

amitayk committed:
build

amitayk committed:
Merge branch 'dev' of https://github.com/di-sukharev/opencommit into allowForCommitPrefix

amyu98 commented:
@di-sukharev 
All set up and ready to go

amitayk committed:
build

amyu98 commented:
@di-sukharev Do we want anything else done?

czlowiek488 commented:
How is this PR going? @di-sukharev Do we need something more?

mreduar commented:
@di-sukharev

di-sukharev commented:
going to merge now, finally, sorry, was reallyy busy :)

btw we won GitHub 2023 hackathon with the OpenCommit, congrats <3

di-sukharev commented on src/commands/config.ts:
```diff
@@ -232,3 +244,36 @@ export const configCommand = command(
+  }
+  
+  if (!isRegex(prefix)) {
+    outro('This Regex format is not supported, please use the classic /*/ foramt')
```
```suggestion
    outro('This Regex format is not supported, please use the classic /*/ format')
```

di-sukharev committed:
Update src/commands/config.ts

di-sukharev commented on src/api.ts:
```diff
@@ -108,4 +117,59 @@ export const getOpenCommitLatestVersion = async (): Promise<
+
+function getCurrentGitBranch(): string | undefined {
+  try {
+    const branchName = execaCommandSync('git symbolic-ref --short HEAD').stdout.trim()
```
is this going to add the full branch name in the commit like: `#234-some-branch-name commit message text goes here`?

amyu98 replied:
No, it will take a regex out of the name. See my first commit on this PR.
Very cool to hear about the hackathon!

di-sukharev commented:
added 1 comment, could you also pls take a look at [this PR](https://github.com/di-sukharev/opencommit/pull/208), isn't it what you want here, but less code and more agile?

amyu98 commented:
Hmm the other PR might do the trick 😓 @di-sukharev

mreduar commented:
But the other PR does not extract the number automatically. You would have to pass the placeholder each time a commit is made.

inguelcartful commented:
Any news on getting this merged?

di-sukharev commented:
managing to merge this today, i need to fix the conflicts first

di-sukharev commented:
@amyu98 if you have time today, maybe it would be faster for you to solve the conflicts :) 

im managing other PRs at the same time, if you chill today as we all should—will fix myself, not sure if i have so much time, there is lots of work to do..

amyu98 commented:
I'll do my best to find the time

On Sun, 3 Sept 2023, 9:08 Sukharev, ***@***.***> wrote:

> @amyu98 <https://github.com/amyu98> if you have time today, maybe it
> would be faster for you to solve the conflicts :)
>
> im managing other PRs at the same time, if you chill today as we all
> should—will fix myself, not sure if i have so much time, there is lots of
> work to do..
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/di-sukharev/opencommit/pull/160#issuecomment-1704021501>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANXHFIFTKLOY4M65ZU5RKZTXYQNEXANCNFSM6AAAAAAXWAW7QA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

di-sukharev commented:
Статический prefix уже можно выразить текущим OCO_MESSAGE_TEMPLATE_PLACEHOLDER и аргументом-шаблоном вокруг $msg. Ветка конфликтует, затрагивает старый API и содержит большие generated bundles. Если branch-regex prefix всё ещё нужен, лучше отдельный свежий issue/PR с безопасным парсингом regex и tests; этот PR рекомендую закрыть.
